controller/object_store_controller.go

```diff
@@ -0,0 +1,1170 @@
+			return
+		}
+		svc, ok = deleted.Obj.(*corev1.Service)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object %#v", deleted.Obj))
+			return
+		}
+
+	}
+
+	// only consider services within the longhorn namespace and which have an
+	// owner. All others can not be related to an object store.
+	if svc.Namespace != osc.namespace || len(svc.ObjectMeta.OwnerReferences) < 1 {
+		return
+	}
+	storeName := svc.ObjectMeta.OwnerReferences[0].Name
+	store, err := osc.ds.GetObjectStoreRO(storeName)
+	if err != nil {
+		return // service has owner reference, but is not owned by an object store
+	}
+	key, err := cache.MetaNamespaceKeyFunc(store)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to get key for object store %v: %v", storeName, err))
+		return
+	}
+	osc.queue.Add(key)
+}
+
+func (osc *ObjectStoreController) enqueuePVC(obj interface{}) {
+	pvc, ok := obj.(*corev1.PersistentVolumeClaim)
+	if !ok {
+		deleted, ok := obj.(cache.DeletedFinalStateUnknown)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("received unexpected obj: %#v", obj))
+			return
+		}
+		pvc, ok = deleted.Obj.(*corev1.PersistentVolumeClaim)
+		if !ok {
+			utilruntime.HandleError(fmt.Errorf("DeletedFinalStateUnknown contained invalid object %#v", deleted.Obj))
+			return
+		}
+
+	}
+
+	// only consider PVCs within the longhorn namespace and which have an
+	// owner. All others can not be related to an object store.
+	if pvc.Namespace != osc.namespace || len(pvc.ObjectMeta.OwnerReferences) < 1 {
+		return
+	}
+	storeName := pvc.ObjectMeta.OwnerReferences[0].Name
+	store, err := osc.ds.GetObjectStoreRO(storeName)
+	if err != nil {
+		return //  pvc has owner reference, but is not owned by an object store
+	}
+	key, err := cache.MetaNamespaceKeyFunc(store)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("failed to get key for object store %v: %v", storeName, err))
+		return
+	}
+	osc.queue.Add(key)
+}
+
+func (osc *ObjectStoreController) reconcile(key string) error {
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	store, err := osc.ds.GetObjectStore(name)
+	if err != nil {
+		if datastore.ErrorIsNotFound(err) {
+			return nil // already deleted, nothing to do
+		}
+		return err
+	} else if !osc.isResponsibleFor(store) {
+		return nil
+	}
+
+	existingStore := store.DeepCopy()
+	defer func() {
+		if reflect.DeepEqual(existingStore.Status, store.Status) {
+			return
+		}
+		store, err = osc.ds.UpdateObjectStoreStatus(store)
+	}()
+
+	// handle termination
+	if !store.DeletionTimestamp.IsZero() {
+		if store.Status.State != longhorn.ObjectStoreStateTerminating {
+			logrus.Infof("object store %v is now terminating", store.Name)
+			store.Status.State = longhorn.ObjectStoreStateTerminating
+			return err
+		}
+		return osc.handleTerminating(store)
+	}
+
+	switch store.Status.State {
+	case longhorn.ObjectStoreStateStarting, longhorn.ObjectStoreStateError:
```

Comment:
I think the current implementation doesn't move from `error` to `starting`. it will remain in `error` until all resources are ready again and it moves the state to `running` ref https://github.com/longhorn/longhorn-manager/pull/2136/files#diff-1b27c9e1a0108c88fe7217cc782c0e8e4cfb5592722b30a3bd54ff83691b5002R411

Reply:
In both the `starting` and `error` state, the controller only creates resources that aren't there already. Then it waits until the resources are healthy as best as it can tell. There are no leftovers to clean up since nothing is deleted if the state is `error`

Reply:
Nobody would guarantee there is no leftover when the state becomes `error`. For example, the stuck deployments or some related components being modified unexpectedly. I think it's better to do a cleanup before updating the state from `error` to `starting`

Reply:
There is nothing the controller can do, besides ensuring that all K8s resources exist, to 'fix' an object store. Neither on-disk data or other resources are controlled by the object store controller directly. Therefore, if an object stores becomes stuck in error state, the best course of action is to just ensure that all K8s resources are there and hope the K8s control plane manages to bring up the Pod and route traffic to it correctly.
Besides, it wouldn't be a good idea to let the controller delete resources like volumes based on some possibly transient error, because that can really lead to data loss. If the error state can't be fixed by just ensuring that K8s resources exist, an operator probably must intervene anyways to avoid data loss.

Reply:
OK. For these native resources, let's utilize the Kubernetes recovery mechanism first and see if it works fine.